test/widgets/test_moc.py

```diff
@@ -0,0 +1,191 @@
+            cls.info[cls.index]['State'] = "PLAY"
+
+        elif arg == "-G":
+            if cls.info[cls.index]['State'] == "PLAY":
+                cls.info[cls.index]['State'] = "PAUSE"
+
+            elif cls.info[cls.index]['State'] == "PAUSE":
+                cls.info[cls.index]['State'] = "PLAY"
+
+        elif arg == "-f":
+            cls.index = (cls.index + 1) % len(cls.info)
+
+        elif arg == "-r":
+            cls.index = (cls.index - 1) % len(cls.info)
+
+
+def no_op(*args, **kwargs):
+    pass
+
+
+class FakeWindow:
+    class _NestedWindow:
+        wid = 10
+    window = _NestedWindow()
+
+
+@pytest.fixture
+def patched_moc(fake_qtile, monkeypatch):
+    widget = moc.Moc()
+    MockMocpProcess.reset()
+    monkeypatch.setattr(widget, "call_process", MockMocpProcess.run)
+    monkeypatch.setattr("libqtile.widget.moc.subprocess.Popen", MockMocpProcess.run)
+    fakebar = Bar([widget], 24)
+    fakebar.window = FakeWindow()
+    fakebar.width = 10
+    fakebar.height = 10
+    fakebar.draw = no_op
+    widget._configure(fake_qtile, fakebar)
+    return widget
+
+
+def test_moc_poll_string_formatting(patched_moc):
+
+    # Both artist and song title
+    assert patched_moc.poll() == "♫ Rick Astley - Never Gonna Give You Up"
+
+    # No artist
+    MockMocpProcess.info[0]["Artist"] = ""
+    assert patched_moc.poll() == "♫ Never Gonna Give You Up"
+
+    # No title
+    MockMocpProcess.info[0]["SongTitle"] = ""
+    assert patched_moc.poll() == "♫ rickroll"
+
+
+def test_moc_state_and_colours(patched_moc):
+
+    # Initial poll - playing
+    patched_moc.poll()
+    assert patched_moc.layout.colour == patched_moc.play_color
+
+    # Toggle pause
+    patched_moc.play()
+    patched_moc.poll()
+    assert patched_moc.layout.colour == patched_moc.noplay_color
+
+    # Toggle pause --> playing again
+    patched_moc.play()
+    patched_moc.poll()
+    assert patched_moc.layout.colour == patched_moc.play_color
+
+
+def test_moc_button_presses(manager_nospawn, minimal_conf_noscreen, monkeypatch):
+
+    # This needs to be patched before initialising the widgets as mouse callbacks
+    # bind subprocess.Popen.
+    monkeypatch.setattr("subprocess.Popen", MockMocpProcess.run)
+    mocwidget = moc.Moc(update_interval=0.05)
+    MockMocpProcess.reset()
+    monkeypatch.setattr(mocwidget, "call_process", MockMocpProcess.run)
+    monkeypatch.setattr("libqtile.widget.moc.subprocess.Popen", MockMocpProcess.run)
+
+    config = minimal_conf_noscreen
+    config.screens = [
+        libqtile.config.Screen(
+            top=libqtile.bar.Bar([mocwidget], 10)
+        )
+    ]
+
+    manager_nospawn.start(config)
+
+    # When started, we have the first item playing
+    topbar = manager_nospawn.c.bar["top"]
+    assert topbar.info()["widgets"][0]["text"] == "♫ Rick Astley - Never Gonna Give You Up"
+
+    # Trigger next item and wait for update poll
+    topbar.fake_button_press(0, "top", 0, 0, button=4)
+    sleep(0.06)
```

Comment:
Not a big fan of using sleep here but I don't think I can force a `poll` when the widget is handled by `manager_nospawn`. `ThreadPollText` doesn't have a `cmd_poll` to be able to force this.

Reply:
Would `manager_nospawn.c.sync()` not work?

Reply:
Oh actually I misunderstood. Maybe it won't. Perhaps using `manager_nospawn.c.eval()`?

Reply:
I was hoping to use `manag_nospawn.c.widget["moc"].function()` but don't know what to pass as an argument as I can't seem to access the widget's internals once it's inside the manager.

Alternatively, I just as a `cmd_poll` to the base widget class which would allow users to force a poll whenever they wanted.

Reply:
It seems all command objects have `cmd_eval` from the base, so you could do `manager_nospawn.c.widget["moc"].eval('self.poll()')`

Reply:
Interesting. I didn't think that would work as the `eval` wasn't called a method of the widget and so wouldn't pass `self` but it looks like it does work.

I can change this to: `success, result = manager_nospawn.c.widget["moc"].eval('self.poll()')` and then test `result`.

Thank you.